File: app/src/main/java/br/com/disapps/homepet/ui/hotels/HotelsPresenter.java
package br.com.disapps.homepet.ui.hotels;

import com.hannesdorfmann.mosby3.mvp.MvpBasePresenter;

import java.util.List;

import br.com.disapps.homepet.data.cache.HotelRepository;
import br.com.disapps.homepet.data.model.Hotel;
import br.com.disapps.homepet.data.ws.response.ListHotelResponse;
import br.com.disapps.homepet.util.rx.IErrorHandlerHelper;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by diefferson.santos on 23/08/17.
 */

public class HotelsPresenter extends MvpBasePresenter<IHotelsView> {

    private final HotelRepository mHotelRepository;
    private final CompositeDisposable disposables = new CompositeDisposable();

    public HotelsPresenter(HotelRepository hotelRepository){
        mHotelRepository = hotelRepository;
    }

    public void loadHoteis(){

        if(isViewAttached()){
            getView().showLoading(false);
        }

        disposables.add(mHotelRepository.getHoteis(true)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(new DisposableObserver<ListHotelResponse>() {
                    @Override
                    public void onNext(ListHotelResponse response) {

                        List<Hotel> hoteis = response.getContent();

                        if(isViewAttached()){
                            getView().fillHotelAdapter(hoteis);
                            getView().dismissLoading();
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        if(isViewAttached()){
                            IErrorHandlerHelper.defaultErrorResolver(HotelsPresenter.this.getView(), e);
                            getView().dismissLoading();
                        }
                    }

                    @Override
                    public void onComplete() {

                    }
                }));
    }

    @Override
    public void detachView(boolean retainInstance) {
        disposables.clear();
    }

}

File: app/src/main/java/br/com/disapps/homepet/ui/hotels/HotelsPresenter.kt
package br.com.disapps.homepet.ui.hotels

import com.hannesdorfmann.mosby3.mvp.MvpBasePresenter

import br.com.disapps.homepet.data.cache.HotelRepository
import br.com.disapps.homepet.data.ws.request.HotelsRequest
import br.com.disapps.homepet.data.ws.response.ApiListResponse
import br.com.disapps.homepet.util.rx.IErrorHandlerHelper
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

/**
 * Created by diefferson.santos on 23/08/17.
 */

class HotelsPresenter(private val mHotelRepository: HotelRepository) : MvpBasePresenter<IHotelsView>() {

    private val disposables = CompositeDisposable()

    fun loadHoteis(request: HotelsRequest) {

        if (isViewAttached) {
            view.showLoading(false)
        }

        disposables.add(mHotelRepository.getHotels(true, request)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableObserver<ApiListResponse.ListHotelResponse>() {
                    override fun onNext(response: ApiListResponse.ListHotelResponse) {

                        val hoteis = response.content

                        if (isViewAttached) {
                            view.fillHotelAdapter(hoteis!!)
                            view.dismissLoading()
                        }
                    }

                    override fun onError(e: Throwable) {
                        if (isViewAttached) {
                            IErrorHandlerHelper.defaultErrorResolver(this@HotelsPresenter.view, e)
                            view.dismissLoading()
                        }
                    }

                    override fun onComplete() {

                    }
                }))
    }

    override fun detachView(retainInstance: Boolean) {
        disposables.clear()
    }
}
